hibernate-core/src/main/java/org/hibernate/boot/model/internal/BinderHelper.java

```diff
@@ -334,7 +336,12 @@ private static Property makeSyntheticComponentProperty(
 		result.setInsertable( false );
 		result.setValue( embeddedComponent );
 		result.setPropertyAccessorName( "embedded" );
-		ownerEntity.addProperty( result );
+		if ( ownerEntity instanceof JoinedSubclass ) {
+			ownerEntity.addProperty( result );
+		}
+		else {
+			persistentClassOrJoin.addProperty( result );
+		}
```

Comment:
Sure, this is actually the one change I was not 100% sure about. I noticed the object where the synthetic `Component` property for the `JoinColumn` was added was often wrong when considering where the referenced column was actually located.

In most simple cases, the column owner (`persistentClassOrJoin`) and `ownerEntity` were the same so this effectively changes nothing.

Even when the column owner and the entity were different, though, the property was always added to the entity. This was causing the column to be always associated with the primary table, and in cases like `SecondaryTable` joins we had no way of knowing the column was actually referencing another table. This change was needed to later be able to get the correct table reference in `ToOneAttributeMapping#createTableGroupJoin`.

The only case where `persistentClassOrJoin` and `ownerEntity` were different and the referenced column was actually in the `ownerEntity`'s side I encountered when inheritance with `InheritanceType.JOINED` was involved, thus the `instanceof` check to add the property to the correct persistent class.

Reply:
It's a bit hard to follow all that without some example mappings to see what the individual cases are.

But it seems like if I stipulate that what you're saying is correct, the logic there is still a bit weird.

Wouldn't it be more natural to write:

```java
if ( persistentClassOrJoin instanceof PersistentClass ) {
 	ownerEntity.addProperty( result );
 }
 else { //must be a join
 	persistentClassOrJoin.addProperty( result );
 }
```

But I'm not sure.

It sounds to me like maybe we need additional tests to exercise all the cases you ran into.

Reply:
Actually maybe you can just give me an example of the mapping which breaks *without* this change?

Reply:
probably it should be
```
if ( persistentClassOrJoin instanceof Join ) {
	persistentClassOrJoin.addProperty( result );
}
else {
	ownerEntity.addProperty( result );
}
```

Reply:
@gavinking 
> Actually maybe you can just give me an example of the mapping which breaks _without_ this change?

Well it doesn't really _break_, but `RefToSecondaryTableTest` mapping of the `Split` entity is the perfect example: without this change the synthetic property is added to the primary table (`Split`) while the referenced column is actually in the secondary one `secondary_split`.

@dreab8 
> what about ...

That should be the same as the new check Gavin proposed, logically. Both look way better than my solution and to me they make sense AFAIU. I'm gonna try to apply them and tests to see what happens.

Reply:
> Well it doesn't really break, but `RefToSecondaryTableTest`'s mapping of the `Split` entity is the perfect example: without this change the synthetic property is added to the primary table (`Split`) while the referenced column is actually in the secondary one `secondary_split`.

Right but that example has nothing to do with `InheritanceType.JOINED`, so I'm not seeing where `JoinedSubclass` enters into this.

Reply:
(In fact that example has no inheritance at all.)

Reply:
> Sorry the inheritance example I encountered was `ManyToManyTest` using `InspectorPrefixes` which is a joined sublcass of `Inspector`.

Ugh. I hate that mess of a package.

Do you think you could create a simpler `RefToJoinedSubclassTest` test case, similar to `RefToSecondaryTableTest`, which demonstrates what breaks when we _don't_ make this change?

Because if there's a problem with references to joined subclasses, then that's actually I guess a *separate bug*, not the same bug I reported.

Reply:
There was no problem with references to joined subclasses. I just needed to add the synthetic property to the `persistentClassOrJoin`, but for joined subclasses this was wrong so I needed the check.

Anyway I've added the new test to be sure and I rewrote the `if` condition like @dreab8 proposed.

Reply:
Ah, sure, sure, now with the updated condition, it's much clearer, actually. I think we're good.

Thank you!!